In [2]:
import requests
import json

# -------------------------
# MODO DE SEGURANÇA (fallback)
# -------------------------
# Se a API não responder, o notebook usa dados locais
# para a atividade continuar funcionando.

DADOS_FALLBACK = {
    "https://swapi.dev/api/people/10/": {
        "name": "Obi-Wan Kenobi",
        "height": "182",
        "mass": "77",
        "eye_color": "blue-gray",
        "gender": "male",
        "url": "https://swapi.dev/api/people/10/"
    },
    "https://swapi.dev/api/people/13/": {
        "name": "Chewbacca",
        "height": "228",
        "mass": "112",
        "eye_color": "blue",
        "gender": "male",
        "url": "https://swapi.dev/api/people/13/"
    },
    "https://swapi.dev/api/planets/2/": {
        "name": "Alderaan",
        "climate": "temperate",
        "terrain": "grasslands, mountains",
        "population": "2000000000",
        "url": "https://swapi.dev/api/planets/2/"
    },
    "https://swapi.dev/api/starships/10/": {
        "name": "Millennium Falcon",
        "model": "YT-1300 light freighter",
        "manufacturer": "Corellian Engineering Corporation",
        "max_atmosphering_speed": "1050",
        "url": "https://swapi.dev/api/starships/10/"
    }
}

class RespostaLocal:
    def __init__(self, status_code, data):
        self.status_code = status_code
        self._data = data
        self.text = json.dumps(data, ensure_ascii=False)

    def json(self):
        return self._data

def buscar_endpoint(url):
    try:
        resposta = requests.get(url, timeout=10)
        if resposta.status_code == 200:
            return resposta, "online"
    except Exception:
        pass

    if url in DADOS_FALLBACK:
        return RespostaLocal(200, DADOS_FALLBACK[url]), "fallback local"

    return RespostaLocal(404, {"detail": "Not found"}), "fallback local"

print("Ambiente preparado com sucesso.")

Ambiente preparado com sucesso.


Identificação do grupo
Integrante 1: Tainan Lima - 101291

1.1 Personagem 12


In [4]:
url = "https://swapi.dev/api/people/12/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Wilhuff Tarkin",
    "height": "180",
    "mass": "unknown",
    "hair_color": "auburn, grey",
    "skin_color": "fair",
    "eye_color": "blue",
    "birth_year": "64BBY",
    "gender": "male",
    "homeworld": "https://swapi.dev/api/planets/21/",
    "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/6/"
    ],
    "species": [],
    "vehicles": [],
    "starships": [],
    "created": "2014-12-10T16:26:56.138000Z",
    "edited": "2014-12-20T21:17:50.330000Z",
    "url": "https://swapi.dev/api/people/12/"
}


Respostas do aluno — Personagem 13

Status code: 200

Nome do personagem: Wilhuff Tarkin

Altura:180

Cor dos olhos: Azul

O JSON parece completo? Sim

1.2 Personagem 16

In [5]:
url = "https://swapi.dev/api/people/16/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Jabba Desilijic Tiure",
    "height": "175",
    "mass": "1,358",
    "hair_color": "n/a",
    "skin_color": "green-tan, brown",
    "eye_color": "orange",
    "birth_year": "600BBY",
    "gender": "hermaphrodite",
    "homeworld": "https://swapi.dev/api/planets/24/",
    "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/3/",
        "https://swapi.dev/api/films/4/"
    ],
    "species": [
        "https://swapi.dev/api/species/5/"
    ],
    "vehicles": [],
    "starships": [],
    "created": "2014-12-10T17:11:31.638000Z",
    "edited": "2014-12-20T21:17:50.338000Z",
    "url": "https://swapi.dev/api/people/16/"
}


Respostas do aluno — Personagem 16

Status code: 200

Nome do personagem: Jabba Desilijic Tiure

Altura: 175

Cor dos olhos: Orange

O JSON parece completo? Sim


1.3 Planeta 5



In [6]:
url = "https://swapi.dev/api/planets/5/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Dagobah",
    "rotation_period": "23",
    "orbital_period": "341",
    "diameter": "8900",
    "climate": "murky",
    "gravity": "N/A",
    "terrain": "swamp, jungles",
    "surface_water": "8",
    "population": "unknown",
    "residents": [],
    "films": [
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/",
        "https://swapi.dev/api/films/6/"
    ],
    "created": "2014-12-10T11:42:22.590000Z",
    "edited": "2014-12-20T20:58:18.425000Z",
    "url": "https://swapi.dev/api/planets/5/"
}


Respostas do aluno — Planeta 5

Status code: 200

Nome do planeta: Dagobah

Clima: Murky

Terreno: swamp, jungles

População: Desconhecido

O JSON parece completo? Sim

1.4 Nave 12

In [8]:
url = "https://swapi.dev/api/starships/12/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "X-wing",
    "model": "T-65 X-wing",
    "manufacturer": "Incom Corporation",
    "cost_in_credits": "149999",
    "length": "12.5",
    "max_atmosphering_speed": "1050",
    "crew": "1",
    "passengers": "0",
    "cargo_capacity": "110",
    "consumables": "1 week",
    "hyperdrive_rating": "1.0",
    "MGLT": "100",
    "starship_class": "Starfighter",
    "pilots": [
        "https://swapi.dev/api/people/1/",
        "https://swapi.dev/api/people/9/",
        "https://swapi.dev/api/people/18/",
        "https://swapi.dev/api/people/19/"
    ],
    "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/"
    ],
    "created": "2014-12-12T11:19:05.340000Z",
    "edited": "2014-12-20T21:23:49.886000Z",
    "url": "https://swapi.dev/api/starships/12/"
}


Respostas do aluno — Nave 12

Status code: 200

Nome da nave: X-wing

Modelo: T-65 X-wing

Fabricante: Incom Corporation

Velocidade máxima em atmosfera: 1050

O JSON parece completo? Sim

2.1 Extraindo campos do personagem

In [9]:
url = "https://swapi.dev/api/people/15/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Nome:", dados["name"])
print("Altura:", dados["height"])
print("Cor dos olhos:", dados["eye_color"])
print("Gênero:", dados["gender"])

Nome: Greedo
Altura: 173
Cor dos olhos: black
Gênero: male


Perguntas 2.1
1. Qual campo foi usado para mostrar o nome? Name

2. Qual campo foi usado para mostrar a altura? Height

3. Se você fosse QA, quais desses campos validaria primeiro? Name porque identifica o personagem.

2.2 Extraindo campos do planeta

In [11]:
url = "https://swapi.dev/api/planets/6/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Nome do planeta:", dados["name"])
print("Clima:", dados["climate"])
print("Terreno:", dados["terrain"])
print("População:", dados["population"])

Nome do planeta: Bespin
Clima: temperate
Terreno: gas giant
População: 6000000


Perguntas 2.2

O campo população parece numérico? Sim.

O clima foi retornado corretamente? Sim

Se você fosse QA, o que validaria nesse planeta? Name, porque identifica o planeta.

Parte 3 — Pensando como QA

In [12]:
url = "https://swapi.dev/api/people/15/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Validação 1 - status 200:", response.status_code == 200)
print("Validação 2 - campo 'name' existe:", "name" in dados)
print("Validação 3 - nome não está vazio:", dados["name"] != "")
print("Validação 4 - campo 'height' existe:", "height" in dados)

Validação 1 - status 200: True
Validação 2 - campo 'name' existe: True
Validação 3 - nome não está vazio: True
Validação 4 - campo 'height' existe: True


PARTE 3

Explique com suas palavras


O que a validação 1 verifica? O STATUS

O que a validação 2 verifica? o nome

O que a validação 3 verifica? se o campo nome não está vazio.

O que a validação 4 verifica? se existe altura.

3.2 Complete as validações

In [13]:
url = "https://swapi.dev/api/starships/12/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Status 200:", response.status_code == 200)

# Os exemplos abaixo já estão corretos e executáveis.
# Se o professor pedir, vocês podem alterar para validar outros campos.
print("Campo 'name' existe:", "name" in dados)
print("Campo 'model' existe:", "model" in dados)
print("Nome não está vazio:", dados["name"] != "")


Status 200: True
Campo 'name' existe: True
Campo 'model' existe: True
Nome não está vazio: True


Resposta 3.2

Status 200: garante que a requisição para a API foi bem-sucedida.

Campo "name" existe: verifica se o campo do nome da nave está presente na resposta.

Campo "model" existe: verifica se o campo do modelo da nave foi retornado pela API.

Nome não está vazio: garante que o campo name tem um valor válido e não está vazio.

Parte 4 — Testes automatizados com assert

Tarefa: execute o bloco. Depois, expliquem o que cada assert valida e, se quiserem, alterem o endpoint para outro personagem.

In [14]:
url = "https://swapi.dev/api/people/14/"
response, origem = buscar_endpoint(url)
dados = response.json()

assert response.status_code == 200, "Erro: status inesperado"
assert "name" in dados, "Erro: campo name não encontrado"
assert dados["name"] != "", "Erro: nome vazio"

print("Todos os testes do personagem passaram!")

Todos os testes do personagem passaram!


Respostas 4

Assert 1: verifica se o status code é 200, garantindo que a requisição foi bem-sucedida.

Assert 2: verifica se o campo name existe no JSON retornado pela API.

Assert 3: verifica se o campo name não está vazio, garantindo que o personagem tem um nome válido.

4.2 Assert na nave
Tarefa: execute o bloco. Depois, descrevam qual campo foi validado em cada assert e proponham mais 1 validação extra em Markdown.

In [15]:
url = "https://swapi.dev/api/starships/12/"
response, origem = buscar_endpoint(url)
dados = response.json()

assert response.status_code == 200, "Erro: status inesperado"
assert "manufacturer" in dados, "Erro: fabricante não encontrado"
assert dados["model"] != "", "Erro: modelo vazio"

print("Todos os testes da nave passaram!")


Todos os testes da nave passaram!


Resposta 4.2

Assert 1: valida se o status code é 200, indicando que a requisição foi bem-sucedida.

Assert 2: valida se o campo manufacturer (fabricante) existe no JSON da nave.

Assert 3: valida se o campo model não está vazio.

Parte 5 — Cenário negativo

In [16]:
url = "https://swapi.dev/api/people/90909/"
response, origem = buscar_endpoint(url)

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print("Texto retornado:")
print(response.text)

Origem dos dados: fallback local
Status Code: 404
Texto retornado:
{"detail": "Not found"}


Resposta 5

Qual status code foi retornado? 404

Esse comportamento está correto? Sim, porque o recurso solicitado não existe na API.

Por que um QA deve testar cenários negativos? Para garantir que o sistema responda corretamente quando ocorre um erro ou quando um recurso inexistente é solicitado.

Parte 6 — Mini desafio final
Agora você vai alterar apenas o endpoint da função abaixo e gerar um pequeno resumo.

In [17]:
def resumo_recurso(url):
    response, origem = buscar_endpoint(url)
    print("Origem dos dados:", origem)
    print("Status Code:", response.status_code)

    if response.status_code != 200:
        print("Recurso não encontrado.")
        return

    dados = response.json()

    for chave, valor in dados.items():
        print(f"{chave}: {valor}")

# ALTERE O ENDPOINT ABAIXO PARA TESTAR OUTRO RECURSO
resumo_recurso("https://swapi.dev/api/planets/4/")

Origem dos dados: online
Status Code: 200
name: Hoth
rotation_period: 23
orbital_period: 549
diameter: 7200
climate: frozen
gravity: 1.1 standard
terrain: tundra, ice caves, mountain ranges
surface_water: 100
population: unknown
residents: []
films: ['https://swapi.dev/api/films/2/']
created: 2014-12-10T11:39:13.934000Z
edited: 2014-12-20T20:58:18.423000Z
url: https://swapi.dev/api/planets/4/


Respostas finais

Qual endpoint você testou no desafio? https://swapi.dev/api/planets/4/

Qual foi o status code? 200

Qual campo chamou mais a sua atenção? climate (frozen)

Se você fosse QA, quais 3 validações faria nesse recurso?

Verificar se o status code é 200.

Verificar se o campo name existe no JSON.

Verificar se o campo climate não está vazio.